SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/InterestManager.cpp

```diff
@@ -0,0 +1,615 @@
+
+namespace SpatialGDK
+{
+FTagComponentStorage::FTagComponentStorage(Worker_ComponentId InComponentId)
+	: TagComponentId(InComponentId)
+{
+	Components.Add(TagComponentId);
+}
+
+void FTagComponentStorage::OnComponentAdded(Worker_EntityId EntityId, Worker_ComponentId ComponentId, Schema_ComponentData* Data)
+{
+	if (ensureAlways(TagComponentId == ComponentId))
+	{
+		TaggedObjects.Add(EntityId);
+		Modified.Add(EntityId);
+	}
+}
+
+void FTagComponentStorage::OnRemoved(Worker_EntityId EntityId)
+{
+	TaggedObjects.Remove(EntityId);
+	Modified.Remove(EntityId);
+}
+
+void FTagComponentStorage::OnUpdate(Worker_EntityId EntityId, Worker_ComponentId InComponentId, Schema_ComponentUpdate* Update)
+{
+	if (!ensure(false))
+	{
+		return;
+	}
+	Modified.Add(EntityId);
+}
+
+struct FInterestManager::FBroadphaseImpl
+{
+	FBroadphaseImpl() {}
+
+	void Add(uint32 Slot, FVector Position) {}
+
+	void Add(uint32 Slot, FBox2D Box) {}
+
+	void Update(uint32 Slot, FVector Position) {}
+
+	void SwapEntries(int32 Slot1, int32 Slot2) {}
+
+	void Remove(uint32 Slot) {}
+
+	void ComputeBoxVisibility(const TArray<FBox2D>& Regions, uint64* OutVisibility) {}
+};
+
+FInterestManager::~FInterestManager() = default;
+
+FInterestManager::FInterestManager(InterestFactory& InInterestF, FSpatialPositionStorage& InPositions,
+								   FAlwaysRelevantStorage& InAlwaysRelevant, FServerAlwaysRelevantStorage& InServerAlwaysRelevant)
+	: InterestF(InInterestF)
+	, Positions(InPositions)
+	, AlwaysRelevant(InAlwaysRelevant)
+	, ServerAlwaysRelevant(InServerAlwaysRelevant)
+{
+	SlotMap.Reserve(1 << 16);
+	Entities.Reserve(1 << 16);
+	EntityPosition.Reserve(1 << 16);
+	EntityFlags.Reserve(1 << 16);
+	Visibility.Reserve(1 << 16);
+	ActiveTimestamp.Reserve(1 << 16);
+}
+
+void FInterestManager::Advance(const TSet<Worker_EntityId_Key>& DeletedEntities)
+{
+	for (Worker_EntityId Entity : DeletedEntities)
+	{
+		if (int32* SlotPtr = SlotMap.Find(Entity))
+		{
+			Remove(*SlotPtr);
+			if (Broadphase)
+			{
+				Broadphase->Remove(*SlotPtr);
+			}
+			SlotMap.Remove(Entity);
+		}
+	}
+
+	for (Worker_EntityId Entity : Positions.GetModifiedEntities())
+	{
+		int32 Slot;
+		int32* SlotPtr = SlotMap.Find(Entity);
+		const FVector& Pos = Positions.GetPositions()[Entity];
+		if (SlotPtr != nullptr)
+		{
+			Slot = *SlotPtr;
+			if (Broadphase)
+			{
+				Broadphase->Update(Slot, Pos);
+			}
+		}
+		else
+		{
+			// Assume these flags are available immediately and never change.
+			uint32 Flags = 0;
+			if (AlwaysRelevant.GetObjects().Contains(Entity))
+			{
+				Flags |= AlwaysRelevantFlag;
+			}
+			if (ServerAlwaysRelevant.GetObjects().Contains(Entity))
+			{
+				Flags |= ServerAlwaysRelevantFlag;
+			}
+			Slot = Allocate();
+			SlotMap.Add(Entity, Slot);
+			Entities[Slot] = Entity;
+			EntityFlags[Slot] = Flags;
+			if (Broadphase)
+			{
+				Broadphase->Add(Slot, Pos);
+			}
+		}
+
+		EntityPosition[Slot] = FVector2D(Pos.X, Pos.Y);
```

Comment:
Assume the EntityPosition and EntityPositionX structs are performance optimization?

Reply:
Yes - https://en.wikipedia.org/wiki/AoS_and_SoA